repos/system_upgrade/supplements/actors/checkreboothygiene/libraries/checkreboothygiene.py

```diff
@@ -0,0 +1,126 @@
+    except (OSError, CalledProcessError):
+        api.current_logger().warning("Failed to query grubby for default kernel", exc_info=True)
+        return
+
+    if running_kernel_version != default_kernel_version:
+        summary = (
+            "This host is not running the default kernel version currently configured in the bootloader. "
+            "Maybe a new kernel version was installed without then rebooting the host? "
+            "Policy requires the host to be rebooted before going forward with the upgrade."
+        )
+
+        kernel_versions_related = [
+            reporting.RelatedResource("default_kernel_version", default_kernel_version),
+            reporting.RelatedResource("running_kernel_version", running_kernel_version),
+        ]
+
+        report = [
+            reporting.Title("Mismatched kernel versions detected"),
+            reporting.Summary(summary),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([reporting.Groups.SANITY, reporting.Groups.INHIBITOR]),
+            reporting.Remediation(hint="Please reboot the host machine."),
+        ] + kernel_versions_related
+        reporting.create_report(report)
+
+
+def check_modified_boot_files():
+    boot_files = []
+    for root, _, files in os.walk("/boot"):
+        for file in files:
+            file_path = os.path.join(root, file)
+            if os.path.isfile(file_path):
+                boot_files.append(file_path)
+
+    modified_boot_files = []
+    uptime_seconds = _get_uptime()
+    for file in boot_files:
+        if os.stat(file).st_mtime > time.time() - uptime_seconds:
+            modified_boot_files.append(file)
+
+    if modified_boot_files:
+        summary = (
+            "Some files on the /boot partition have been modified since the last reboot. "
+            "To reduce the risk of boot issues related to changes made since the last reboot, "
+            "policy requires the host to be rebooted before going forward with the upgrade."
+        )
+
+        boot_files_related = [reporting.RelatedResource("file", f) for f in modified_boot_files]
+
+        report = [
+            reporting.Title("Modified files under /boot detected"),
+            reporting.Summary(summary),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([reporting.Groups.SANITY, reporting.Groups.INHIBITOR]),
+            reporting.Remediation(hint="Please reboot the host machine."),
+        ] + boot_files_related
+        reporting.create_report(report)
+
+
+def _get_uptime():
+    """
+    Get system uptime in seconds.
+
+    :returns: Uptime in seconds
+    :rtype: float
+    """
+    with open("/proc/uptime", "r") as f:
+        uptime_seconds = float(f.readline().split()[0])
+    return uptime_seconds
+
+
+# NOTE: got from repos/system_upgrade/common/actors/ipuworkflowconfig/libraries/ipuworkflowconfig.py
+#       should this be moved to a common library?
+def _get_booted_kernel():
```

Comment:
Move this to common library (used by ipuworkflowconfig actor)? Or create actor which produces the booted kernel version...

Reply:
> @Wecros and @pirat89 , thank you for working on this PR. Here's some example verbiage adapted from an OpenSCAP check I developed for these requirements. You can parse these to pad out the texts for rationale, summary, remediation hint, etc. The idea is that an organization would have a policy that OS maintenance that will require a reboot shall not be undertaken if the uptime is greater than some threshold.
> 
> > For the README: The CheckRebootHygiene actor will report an inhibitor in case of conditions that may be prohibited by an organization's reboot hygiene policy. For example, the actor will report inhibitor risk if the host uptime is greater than a maximum defined by the policy, if the running kernel version does not match the default kernel version configured in the bootloader, or if any files are found under /boot that have been modified since the last reboot.
> 
> > This host has not been rebooted for over {{max_allowed_uptime_text}}. To reduce the risk of boot issues related to any changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade.
> 
> > This host is not running the default kernel version currently configured in the bootloader. Maybe a new kernel version was installed without then rebooting the host? Policy requires the host to be rebooted before going forward with the upgrade.
> > Bootloader default: {{default_kernel_version}}
> > Running kernel: {{uname_r_version}}
> 
> > Files on the /boot partition listed below have been modified since the last reboot. To reduce the risk of boot issues related to changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade.
> > {{list_of_files_modified}}

Thanks for providing more info. Something like that is good to have in the description of the actor so people know exactly what to expect from the actor and whether they want to apply it.

Reply:
> Move this to common library (used by ipuworkflowconfig actor)? Or create actor which produces the booted kernel version...

let's discuss that after the sync mtg.